cranelift-codegen/src/redundant_reload_remover.rs

```diff
@@ -0,0 +1,917 @@
+
+impl RedundantReloadRemover {
+    // Push a clone of the top-of-stack ProcessingStackElem.  This will be used to process exactly
+    // one Ebb.  The diversions are created new, rather than cloned, to reflect the fact
+    // that diversions are local to each Ebb.
+    fn processing_stack_push(&mut self, n_regunits: u16, cursor: CursorPosition) {
+        let avail_env = if self.processing_stack.is_empty() {
+            AvailEnv::new(n_regunits as usize)
+        } else {
+            let tos = self.processing_stack.len() - 1;
+            self.processing_stack[tos].avail_env.clone()
+        };
+        let diversions = RegDiversions::new();
+        self.processing_stack.push(ProcessingStackElem {
+            avail_env,
+            cursor,
+            diversions,
+        });
+    }
+
+    // This pushes the node `dst` onto the processing stack, and sets up the new
+    // ProcessingStackElem accordingly.  But it does all that only if `dst` is part of the current
+    // tree *and* we haven't yet visited it.
+    fn processing_stack_maybe_push(&mut self, n_regunits: u16, dst: Ebb) {
+        if self.nodes_in_tree.contains(dst) && !self.nodes_already_visited.contains(dst) {
+            if !self.processing_stack.is_empty() {
+                // If this isn't the outermost node in the tree (that is, the root), then it must
+                // have exactly one predecessor.  Nodes with no predecessors are dead and not
+                // incorporated in any tree.  Nodes with two or more predecessors are the root of
+                // some other tree, and visiting them as if they were part of the current tree
+                // would be a serious error.
+                debug_assert!(self.npreds_per_ebb[dst.as_u32() as usize] == 1);
+            }
+            self.processing_stack_push(n_regunits, CursorPosition::Before(dst));
+            self.nodes_already_visited.insert(dst);
+        }
+    }
+}
+
+impl<'a> Context<'a> {
+    // Perform redundant-reload removal on the tree shaped region of graph defined by `root` and
+    // `self.state.nodes_in_tree`.
+    fn process_tree(&mut self, root: Ebb) {
+        debug_assert!(self.state.nodes_in_tree.contains(root));
+        debug_assert!(self.state.processing_stack.is_empty());
+        debug_assert!(self.state.nodes_already_visited.is_empty());
+
+        // Create the initial work item
+        self.state
+            .processing_stack_maybe_push(self.n_regunits, root);
+
+        while !self.state.processing_stack.is_empty() {
+            // It is somewhat ridiculous to construct a whole new FuncCursor just so we can do
+            // next_inst() on it once, and then copy the resulting position back out.  Is there a
+            // better way?  All attempts to include FuncCursor directly in ProcessingStackElem
+            // lead to intractable difficulties with the borrow checker.
+            let tos = self.state.processing_stack.len() - 1;
```

Comment:
I changed to using last_mut() in as many places as I could.  But changing this one leads to difficulties with the borrow checker, because we'd want to keep the value borrowed by last_mut() alive across the call to visit_inst(), but that call requires a mutable borrow of the whole RedundantReloadRemover state.